### DISCUSSION TITLE
Using a p:commandLink in a Datatable to open a popup Dialog

### DISCUSSION THREAD

**Discussion by xidbhonaker** [primefaces/community#847] [PrimeFaces] [answered]
Using PrimeFaces 13.0, and I've been trying to add a link to a datatable to popup more information, but everything I've tried isn't getting the row's data into the dialog.

```
	      <p:dataTable border="0" 
			value="#{table_list.packageVersions}" 
			var="table_list_vers"** 
			rowKey="#{table_list_vers.versionKey}"
			showGridlines="true"
			stripedRows="true"
			tableStyleClass="ituglib-datatable">
                   ...
	       <p:column id="fileHash" headerText="Hash" width="60">
                 <p:commandLink value="#{table_list_vers.fileHashType}" rendered="#{table_list_vers.hasHashValue}"
	            oncomplete="PF('varHashDlg').show()" update="hashDlg" />
	       </p:column>
                   ...
	      </p:dataTable>
	      <p:dialog header="Hash Value" id="hashDlg" widgetVar="varHashDlg" modal="true" height="100">
		    <h:outputText id="txtHashValue" value="#{table_list_vers.fileHashValue}" />
	      </p:dialog>   

```

2 things: First, this datatable is embedded in another one and in a p:rowExpansion element (along with the p:dialog).  Don't know if that makes a difference.  Second, note that the commandLink is only shown if pertinent values are in the underlying database.

When I click this, the dialog shows, but the outputText is empty (verified using Edge developer mode).  I had used this commandLink in my previous testing and the value shows up, so the bean value 'fileHashValue' is right.  Of course, alert() is not our goal here:

```
            <p:commandLink value="#{table_list_vers.fileHashType}" rendered="#{table_list_vers.hasHashValue}"
	            oncomplete='alert("#{table_list_vers.fileHashValue}")' />

```
Everything I've read online (including various Google searches) implies that the update= parameter will cause the AJAX variable in the outputText value to update with the current row, but nothing explicitly states that.

Should the p:dialog be outside of both tables, inside the main datatable, or inside the embedded datatable?

Any guidance or help would be greatly appreciated!

**Comment by melloware** [comment]
PF 13.0 or 13.0.4 ?  There are many bugs fixed between 13.0 and 13.0.4 so give the latest a try.

**Comment by xidbhonaker** [comment]
13.0.0
I downloaded it on October 2.

**Reply by melloware** [reply]
Try 13.0.4 it's available on Maven Central. Released 2 weeks ago.

**Reply by xidbhonaker** [reply]
Thanks for the guidelines, Melloware team.  I don't use Maven in my projects so will need to download a package.

Where would I find a list of bugs fixed with each release?  Not often a fan of implementing 2-week-old software (even  Open Source software which I understand as having many reviewers prior to release) for production, so I'd like to review where this might have been fixed.  That is, if it's really a bug and not user error (which I was assuming).

Also, if there are critical bugs fixed, I'd still consider using newer releases (which I do for OpenSSL often!)

**Reply by melloware** [reply]
@xidbhonaker you can download the package manually from Maven Central.  And all release notes per version are here:

https://github.com/primefaces/primefaces/releases/tag/13.0.4

That is the 13.0.4 release notes but you can look at 13.0.3, 13.0.2 etc as well.

**Reply by xidbhonaker** [reply]
Thank you very much!  Will start that research later today.

**Comment by xidbhonaker** [comment]
I have tried 13.0.4.  It fails differently.  Along the way I did modify the commandLink with the Dialog, I changed 

`oncomplete="PF('varHashDlg').show()" update="hashDlg"
`

to this:

`oncomplete="PF('hashDlg').show()" update="PF('varHashDlg')" 
`

As I stated above, the commandLink is in an embedded dataTable.  

With that change, on 13.0.0, when I tried to expand the row element in the 'outer' table, the AJAX response from the server is:
```
<partial-response>
 <error>
  <error-name>org.primefaces.expression.ComponentNotFoundException</error-name>
  <error-message><![CDATA[Cannot find component for expression "PF('varHashDlg')" referenced from "ItugTableForm:j_id_e_1_7:0:j_id_e_1_l:0:j_id_e_1_y".]]></error-message>
 </error>
</partial-response>
```

Running the same code on 13.0.4 (only changing the 1 file in the WEB-INF/lib directory), I get the following:

```
<?xml version="1.0" encoding="ISO-8859-1"?>
<partial-response id="j_id__v_0"><changes>
 <update id="j_id__v_0:javax.faces.ViewState:1">
  <![CDATA _(viewstate which I can't interpret)_ =]]>
  </update>
 </changes>
</partial-response>
```

However, in both cases, the row in the outer table does NOT expand. When I replace the commandLink with the one above that has the alert() call, the row does expand, and the alert() works.  So, I am highly suspect that something in the way I've crafted the commandLink to update and open the dialog is wrong.  I tried not using PF() and got errors. And passing the id of the dialog to oncomplete and the widgetVar of the dialog to the update are the only variants that don't give me a component not found error.

If it's pertinent, the alert() version where the expand works has this in the update element:

```
<changes><update id="ItugTableForm:j_id_e_1_7"> and then CDATA.
```

Element j_id_e_1_7  is the id of the rendered div surrounding the outer table, which I would expect would cause the subordinate table to expand under the row I clicked.


This is the current commandLink that doesn't work:

```
        <p:commandLink value="#{table_list_vers.fileHashType}" rendered="#{table_list_vers.hasHashValue}"
	            oncomplete="PF('hashDlg').show()" update="PF('varHashDlg')" />

```

The 2 bean properties in table_list_vers are correct, because the version using alert() works as expected with the value and rendered properties.

Can anyone see what I might have done wrong here?

**Reply by melloware** [reply]
Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.

I have a feeling your issue might be this `rendered="#{table_list_vers.hasHashValue}"`

**Reply by xidbhonaker** [reply]
I'll work on that.  But the rendered= works correctly with this commandLink:

```
            <p:commandLink value="#{table_list_vers.fileHashType}" rendered="#{table_list_vers.hasHashValue}"
	            oncomplete='alert("#{table_list_vers.fileHashType} hash for file #{table_list_vers.fileName} is #{table_list_vers.fileHashValue}")' />

```
The purpose of that is not to show the link at all if certain fields in the underlying SQL table are nul.  The getHasHashValue method in the bean returns a boolean.

**Reply by melloware** [reply]
OK well then I definitely need a runnable reproducer so I can debug it.

**Reply by xidbhonaker** [reply]
Cloning the project as we speak.  Thanks.  I'm noting that the JSF libraries are different from what I'm using.  Also not a Maven expert... which branch should I start with?

I'm using Apache MyFaces JSF  2.3 with Tomcat 8.5...

**Reply by xidbhonaker** [reply]
I apologize for this, but I don't have or use Maven (no mvn executable); I use Eclipse.  I don't see any Eclipse project files in the test code.  Is there a document that gives me a little bit of a head start on this?

I'm concerned that me trying to learn a new (to me, I know) development platform while reproducing the 'problem' will take a long time and possibly not give you a good example.

**Reply by xidbhonaker** [reply]
If you can point me to a working example somewhere in the documentation of how to use p:commandLink to popup a p:Dialog inside a p:datatable, it might help me more.  I just can't find any usable examples... and I know I'm not the first person to want to do this.  More than likely, something in the p:dialog or p:commandLink I've done has a bug.  But they're pretty simple so maybe another user would see my mistake.

**Reply by melloware** [reply]
I use then without issue in datatable which is why I need a reproducer to see what you are doing wrong.

**Reply by xidbhonaker** [reply]
If you can't send me a sample of your cod that does this, and I can't find examples, the best I can do is oversimplify.

The app has been running many years, in the last few months got upgraded to 13.0.  I'm trying to add one column of functionality, using my understanding of PrimeFaces, so below is the minimum, I think.

Where does the p:dialog need to be in relation to the (two) tables?  Should it be inside one or both p.datatable elements? or outside of them? Right now, it's outside (note: this code is entered as text and has not been compiled or validated):

```
 <h:form id="myForm" >
    <!-- some content -->
<p:datatable id-="main" var="table_list" ... >
    <!-- some content -->
   <p:rowExpansion>
   <p:databable  id="embed" var="table_list_vers" ... >
     <!-- some content -->
     <p:column id="hash" headerText="Hash" ... >
        <p:commandLink value="#{table_list_vers.fileHashType}" rendered="#{table_list_vers.hasHashValue}"
	            oncomplete="PF('dial').show()" update="PF('varDial')" />
     </p:column id 
   </p:databable>
 </p:databable>
<p:dialog id="dial" widgetVar="varDial" ... >
  <!-- some content --> 
  <h:outputText id="txtHashValue" value="#{table_list_vers.fileHashValue}" />
/p:dialog>
    <!-- some content -->
</h:form>
```

The code-behind beans include this in the POJO object referenced by table_list_vers:

```
	public Boolean getHasHashValue( ) { Boolean bRet = false; /* some code /* return bRet; }
        public String getFileHashType() { return "SomeValue"; }
        public String getFileHashValue() { return "SomeValue"; }
```

All 3 of these methods work elsewhere in the xhtml when not using p:dialog.

**Reply by melloware** [reply]
So when you say you only use eclipse?  How do you build your software or WAR file?  From eclipse?  Or is this a homework assignment?  Never using Maven in today's Java world seems strange to me.

**Reply by tandraschko** [reply]
PF('varDial') is invalid anyway
PF('...') is just to access a widget by JS

**Reply by xidbhonaker** [reply]
Thanks, Tandraschko, I have made more progress, and am closer (but not working yet.  Currently the commandLink is coded as follows:

```
<p:commandLink value="#{table_list_vers.fileHashType}" rendered="#{table_list_vers.hasHashValue}"
	            oncomplete="PF('varHashDlg').show()" process="@this" update=":ItugTableForm:pnlDlgHash" >
             <f:setPropertyActionListener value="#{table_list_vers}" target="#{table_list.selectedVersion}" />
	         <p:resetInput target=":ItugTableForm:pnlDlgHash" />
	        </p:commandLink> 
```
Dialog still is not finding current data.  Will update this when I get it working.

**Comment by xidbhonaker** [comment]
Multiple ways.  Eclipse has an 'Export' function that creates a fully executable WR file.  But the target platform is not Linux, nor Windows.  It is POSIX-compliant and runs OpenJDK 11 just fine.  And as a long-time enterprise software developer, I like to make sure code can be built on the target platform from source (in this case from a github repository).  So, I have a bash script that uses javac and jar to create the file.

The platform does support Maven, or so I've been told.  But I've been around the business a long time, from old Assembler and COBOL days up until more modern stuff today.  I cut my teeth on Eclipse (for C/C++ first, Java later).  I've never been trained in Maven, and the one time I tried to use it I couldn't get it to work.  That being said, I can usually read a pom file and understand what it's doing.

Most of my work with Open Source teams is in the C/C++ world, though, so taking the time to learn Maven is a lower priority to me.  I'm very effective doing it the way I always have.  I shouldn't need to apologize for that, I've learned that all sorts of my customers, for example, have their own ways to do things, and I try very hard not to tell them that their ways are strange.

This is not a homework assignment, but it is volunteer work for a not-for-profit.  I'm not being paid for any of it, and the organization doesn't make revenue off the site.

**Reply by melloware** [reply]
OK I understand its juts without a runnable reproducer using PrimeFaces Test its impossible for me to guess what is wrong with your code from the snippets above.  But I just tested it locally and using this example: https://www.primefaces.org/showcase/ui/data/datatable/crud.xhtml?jfwid=0b572

I just changed the Edit Button to a CommandLink and the Dialog opens fine.

**Comment by xidbhonaker** [comment]
That helped a lot, I have corrected the oncomplete and update properties of the commandLink, and added the process property.
  I also added the p:resetInput element.   
I put an outputpanel in the dialog, and added 'the responsive property.

The dialog now pops up corrrectly, but the fields in the dialog are still empty.  The developer mode network tab shows this when I click the link:

```
<?xml version="1.0" encoding="ISO-8859-1"?><partial-response id="j_id__v_0"><changes><update id="ItugTableForm:pnlDlgHash"><![CDATA[<div id="ItugTableForm:pnlDlgHash" class="ui-outputpanel ui-widget">
		  <table style="padding:0;" width="520px">
		   <tr><td width="60px" align="right">Package:</td><td width="450px"><span id="ItugTableForm:txtHashPkg"></span>
		    </td></tr>
		   <tr><td width="60px" align="right">Version:</td><td width="450px"><span id="ItugTableForm:txtHashVers"></span>
		    </td></tr>
		   <tr><td width="60px" align="right">Type:</td><td width="450px"><span id="ItugTableForm:txtHashFileType"></span>
		    </td></tr>
		   <tr><td width="60px" align="right"><span id="ItugTableForm:txtHashType"></span>:
		    </td><td width="450px"><span id="ItugTableForm:txtHashValue"></span>
		    </td></tr>
		  </table></div>]]></update><update id="j_id__v_0:javax.faces.ViewState:1"><![CDATA stuff></update></changes></partial-response>

```

I think I need to add the selection property to the datatable, and the setPropertyActionListetener elemnt to the commandLink.  I need to study up on that, haven't needed a selection property before.  This is a task for tomorrow. 

Great example, thanks again!

**Reply by melloware** [reply]
Excellent!

**Reply by xidbhonaker** [reply, marked as answer]
Finally got everything working... my example was made more complex than examples I found, due to having a datatable-in-a-datatable page, wanting to invoke a dialog from the inner table.  Highlights of what worked:

```
<h:form id="MyTableForm" >
 <p:dataTable value="#{mainBean.dataList}" 
              var="table_list"
                <!-- Other useful stuff here -->
              selectionMode="single"
              selection="#{mainBean.selectedRow}">
  <p:column style="width:16px"><p:rowToggler /></p:column>
                <!-- Other columns etc. here -->
  <p:rowExpansion>
   <p:dataTable value="#{table_list.innerRows}" 
                var="table_list_inner" 
                  <!-- Other useful stuff here -->
                selectionMode="single"
                selection="#{table_list.selectedInner}">
              <!-- Other columns etc. here -->
    <p:column id="dlgLink" headerText="Dialog" width="60">
     <p:commandLink value="#{table_list_inner.linkText}" 
                    rendered="#{table_list_inner.hasValue}"
                    oncomplete="PF('varDlg').show()" process="@this"
                    update=":MyTableForm:pnlDlg" >
     <f:setPropertyActionListener value="#{table_list_inner}"
                                  target="#{table_list.selectedInner}" />
     <f:setPropertyActionListener value="#{table_list}"
                                  target="#{mainBean.selectedRow}" />
     <p:resetInput target=":MyTableForm:pnlDlg" />
     </p:commandLink>    
    </p:column>
   </p:dataTable>
  </p:rowExpansion>
 </p:dataTable>
 <p:dialog header="This Dialog" widgetVar="varDlg" modal="true" width="600">
  <p:outputPanel id="pnlDlg" >
   <table style="padding:0;" width="520px">
    <tr><td width="60px" align="right">Main Value:</td><td width="450px">
         <h:outputText id="txtMainValue" style="font-weight : bold;" 
                       value="#{mainBean.selectedRow.mainValue}" />
    </td></tr>
    <tr><td width="60px" align="right">Inner Value:</td><td width="450px">
         <h:outputText id="txtInnerValue" style="font-weight : bold;" 
                       value="#{mainBean.selectedRow.selectedInner.innerValue}" />
    </td></tr>
   </table>
  </p:outputPanel>
 </p:dialog>   
</h:form>

```
The key things that got it working were:

- The 'selectionmode' and 'selection' on the 2 datatables (along with bean methods)
- The 2 setPropertyActionListener elements and the resetInput  element on the commandLink, and
- The use of the selected bean to find the values from the selected row in the dialog.

Thanks, everyone, for your help and patience!
